internal/provider/resource_tfe_workspace_execution_mode.go

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func resourceTFEWorkspaceExecutionMode() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTFEWorkspaceExecutionModeCreate,
+		Update: resourceTFEWorkspaceExecutionModeUpdate,
+		Read:   resourceTFEWorkspaceExecutionModeRead,
+		Delete: resourceTFEWorkspaceExecutionModeDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"agent_pool_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"workspace_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"execution_mode": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						"agent",
+						"local",
+						"remote",
+					},
+					false,
+				),
+			},
+		},
+	}
+}
+
+func resourceTFEWorkspaceExecutionModeCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	executionMode := d.Get("execution_mode").(string)
+	agentPoolID := d.Get("agent_pool_id").(string)
+
+	if executionMode == "agent" && agentPoolID == "" {
+		return fmt.Errorf(`agent_pool_id must be provided when execution_mode is "agent"`)
+	}
+
+	if executionMode != "agent" && agentPoolID != "" {
+		return fmt.Errorf(`execution_mode must be set to "agent" to assign agent_pool_id`)
+	}
```

Comment:
We should be able to move these errors into the plan phase by creating schema validation or, as a last resort, a `CustomizeDiff` handler. It would be a drag for users to have to wait until apply to find out they didn't configure the resource correctly.

Reply:
`ValidateFunc` on execution_mode won't allow me to do a d.Get on agent_pool_id so I went with `CustomizeDiff`